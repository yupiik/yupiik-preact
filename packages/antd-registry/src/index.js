import * as antdIcons from '@ant-design/icons';
import * as antd from 'antd';
import { simpleComponent } from '@yupiik/dynamic/src/component/registry';

const isComponent = it => {
    const c = it.charAt(0);
    return c === c.toUpperCase() && c !== '$' && c !== '_' && !it.endsWith('Props');
};

const antdComponent = nameOrComponent => {
    const impl = simpleComponent(nameOrComponent);
    /* can forward static props - not that helpful since we have 2 HoC wrapping the actual component
       but antd checks if the child is directly an "Item" sometimes so we can probably hack up a proxy there?
    Object
        .keys(nameOrComponent)
        .filter(it => it.startsWith('__ANT_'))
        .forEach(i => impl[i] = nameOrComponent[i]);
    */
    return impl;
}

const forward = obj => Object
    .keys(obj)
    .filter(it => isComponent(it))
    .reduce((a, i) => {
        const wrapper = obj[i];

        a[i] = antdComponent(wrapper);

        // sub-components (Header, Content, Footer, Item, ...)
        Object
            .keys(wrapper)
            .filter(it => isComponent(it))
            .forEach(it => {
                a[`${i}.${it}`] = antdComponent(wrapper[it]);
            });

        return a;
    }, {});

export const AntdRegistry = {
    ...forward(antd),
    ...forward(antdIcons),
};
