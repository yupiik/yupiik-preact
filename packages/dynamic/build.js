const builder = require('../../esbuild.config.base');

const copyUtilities = () => {
    const fs = require('fs');
    if (!fs.existsSync('dist/esbuild')) {
        fs.mkdirSync('dist/esbuild');
    }
    fs.copyFileSync('src/esbuild/shim.js', 'dist/esbuild/shim.js');
};

(async function build() {
    await builder.build();
    copyUtilities();
})();
