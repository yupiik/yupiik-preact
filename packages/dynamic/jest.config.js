module.exports = require('../../jest.config.base');
module.exports = require('../../jest.config.base');
