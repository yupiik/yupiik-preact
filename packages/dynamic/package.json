{
  "name": "@yupiik/dynamic",
  "description": "A simple preact based library to help configuring an UI.",
  "main": "dist/index.js",
  "mainSrc": "src/index.js",
  "author": "Yupiik <contact@yupiik.com>",
  "license": "ISC",
  "keywords": [
    "preact",
    "yupiik"
  ],
  "scripts": {
    "clean": "rm -Rf dist",
    "build": "node build.js",
    "publish": "npm publish --access public",
    "test": "jest",
    "test:watch": "npx jest --watchAll",
    "test:updateSnapshot": "npx jest --updateSnapshot",
    "test:debug": "node --inspect-brk ./node_modules/jest/bin/jest.js --runInBand"
  }
}
