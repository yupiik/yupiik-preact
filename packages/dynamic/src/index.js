import {
    Dynamic,
    ComponentRegistryContext
} from './component/Dynamic';
import {
    FromConfiguratonHoc,
} from './component/FromConfiguratonHoc';

import {
    HtmlRegistry,
} from './component/registry';

export {
    Dynamic,
    ComponentRegistryContext,
    FromConfiguratonHoc,
    HtmlRegistry,
};
