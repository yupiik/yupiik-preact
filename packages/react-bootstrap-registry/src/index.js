import { simpleComponent } from '@yupiik/dynamic/src/component/registry';
import * as reactBootstrap from 'react-bootstrap';

const isComponent = it =>
    it.charAt(0) === it.charAt(0).toUpperCase() &&
    !it.includes('$') &&
    !it.includes('_') &&
    !it.endsWith('Props');

const forward = obj => Object
    .keys(obj)
    .filter(it => isComponent(it))
    .reduce((a, i) => {
        const wrapper = obj[i];

        a[i] = simpleComponent(wrapper);

        // subcomponents
        Object
            .keys(wrapper)
            .filter(it => isComponent(it))
            .forEach(it => {
                a[`${i}.${it}`] = simpleComponent(wrapper[it]);
            });

        return a;
    }, {});

export const ReactBootstrapRegistry = forward(reactBootstrap);
