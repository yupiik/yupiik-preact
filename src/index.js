import {
    Dynamic,
    ComponentRegistryContext
} from './component/Dynamic';

export {
    Dynamic,
    ComponentRegistryContext,
};
