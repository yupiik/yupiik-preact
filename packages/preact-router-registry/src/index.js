import { h } from 'preact';
import { simpleComponent } from '@yupiik/dynamic';
import { Link, Router, route } from 'preact-router';

// avoids page reloading
const NavLink = ({ href, children, ...rest }) => (
    <a href={href || ''} {...rest} onClick={e => {
        e.preventDefault();
        route(href || '');
    }}>
        {children || ''}
    </a>);

export const PreactRouterRegistry = {
    Router: simpleComponent(Router),
    Link: simpleComponent(Link),
    NavLink: simpleComponent(NavLink),
};
