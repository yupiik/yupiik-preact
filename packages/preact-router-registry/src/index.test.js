import { h } from 'preact';
import { render } from '@testing-library/preact';
import { expect } from 'expect';
import { FromConfiguratonHoc } from '../../dynamic/src';
import { PreactRouterRegistry } from './index';

test('Should render preact-router components', () => {
    // statc part (JSON friendly)
    const opts = {
        name: 'Router',
        options: {
            configuration: {
                history: { // just for the test
                    location: {
                        pathname: '/home',
                    },
                    push: function () { },
                    listen: function () { },
                },
                children: [
                    {
                        name: 'div',
                        options: {
                            wrapperProps: {
                                path: '/home',
                            },
                            configuration: {
                                children: [
                                    'home',
                                ],
                            },
                        },
                    },
                    {
                        name: 'div',
                        options: {
                            wrapperProps: {
                                default: true,
                            },
                            configuration: {
                                children: [
                                    'default',
                                ],
                            },
                        },
                    },
                ],
            },
        },
    };

    // rendering
    const { container: { children } } = render(
        <FromConfiguratonHoc
            registry={PreactRouterRegistry}
            options={opts}
        />
    );

    // test
    expect(children).toMatchSnapshot();
});
