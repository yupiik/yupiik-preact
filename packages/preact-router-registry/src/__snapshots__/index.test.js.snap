// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render react-bootstrap components 1`] = `
HTMLCollection [
  <div>
    home
  </div>,
]
`;
